### PR TITLE
Partial InnerNode implementation

### PR DESCRIPTION
**Pull request by AlexandruCihodaru**
### Summary of the PR

* Added search methods for the tree implementation.
* Added balancing methods for the tree implementation.

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and follow
  [the 50/72 git commit rule](https://www.midori-global.com/blog/2018/04/02/git-50-72-rule).
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: src/allocation_engine/interval_tree.rs [hunk 1]
```diff
@@ -185,8 +185,173 @@ struct InnerNode {
...
+
+    /// Returns a readonly reference to the node associated with the `key` or
+    /// None if there is no Node representing an interval that covers the
+    /// searched key. For a key [a, b], this method will return a node with
+    /// a key [c, d] such that c <= a and b <= d.
+    #[allow(dead_code)]
+    fn search_superset(&self, key: &Range) -> Option<&InnerNode> {
+        if self.key.contains(key) {
+            Some(self)
+        } else if key.end < self.key.start && self.left.is_some() {
```
Comment (gsserge) [inline src/allocation_engine/interval_tree.rs:222]: I wonder if we can use the same trick `.as_ref().and_then()` as in `search()` and get rid of unwraps and `else { None }`. Something like
```rust
if self.key.contains(key) {
    Some(self)
} else if key.end < self.key.start {
    self.left.as_ref().and_then(|node| node.search_superset(key))
} else { // seems like the only possibility here is key.start > self.key.end
    self.right.as_ref().and_then(|node| node.search_superset(key))
}
```

File: src/allocation_engine/interval_tree.rs [hunk 2]
```diff
@@ -185,8 +185,173 @@ struct InnerNode {
...
+    }
+
+    /// Rotate the tree such that height difference between subtrees
+    /// is not greater than abs(1).
+    #[allow(dead_code)]
+    fn rotate(self) -> Box<Self> {
+        let l = height(&self.left);
+        let r = height(&self.right);
+
+        match (l as i32) - (r as i32) {
```
Comment (gsserge) [inline src/allocation_engine/interval_tree.rs:240]: Why cast to `i32`?

File: src/allocation_engine/interval_tree.rs [hunk 3]
```diff
@@ -185,8 +185,173 @@ struct InnerNode {
...
+        let (mut new_root, remaining) = r.get_new_root();
+        new_root.left = Some(Box::new(l));
+        new_root.right = remaining;
+        new_root.updated_node()
+    }
+
+    /// Update cached information of the node.
+    /// Please make sure that the cached values of both children are up to date.
+    #[allow(dead_code)]
+    fn update_cached_info(&mut self) {
```
Comment (gsserge) [inline src/allocation_engine/interval_tree.rs:339]: If only the height is cached, then maybe it's better to be more specific with the method's name and call it `update_cached_height`. `info` sounds a bit generic.

File: src/allocation_engine/interval_tree.rs [hunk 4]
```diff
@@ -185,8 +185,173 @@ struct InnerNode {
...
+        }
+    }
+
+    /// Perform a single left rotation on this node.
+    #[allow(dead_code)]
+    fn rotate_left(mut self) -> Box<Self> {
+        let mut new_root = self.right.take().expect("Node is broken");
+        self.right = new_root.left.take();
+        self.update_cached_info();
+        new_root.left = Some(Box::new(self));
```
Comment (gsserge) [inline src/allocation_engine/interval_tree.rs:254]: I'm a bit concerned with the amount of `Box::new()` calls in the `rotate*` methods. Each call is a memory allocation and is relatively expensive. Conceptually, you need to allocate memory only when instantiating a new node. The rotations should be performed by adjusting pointers, there should be no need for more allocations.

Now, I know that Rust is being, well, Rust, and simple things are usually difficult :) But could you please spend some time and investigate if it's possible to perform the rotations without allocations.

**Review by gsserge — DISMISSED** [review]

**Review by studychao — COMMENTED** [review]
Nice work:) So far so good, but I'll take a look at it again next morning.

File: src/allocation_engine/interval_tree.rs [hunk 5]
```diff
@@ -286,4 +451,157 @@ mod tests {
...
+        right_node1.height = 4;
+        right_node2.height = 3;
+        right_node3.height = 2;
+
+        right_node3.right = Some(Box::new(right_node4));
+        right_node2.right = Some(Box::new(right_node3));
+        right_node1.right = Some(Box::new(right_node2));
+        balanced_tree.left = Some(Box::new(left_node1));
+        balanced_tree.right = Some(Box::new(right_node1));
+        balanced_tree = *balanced_tree.delete_root().unwrap();
```
[the remark this answers is not in the document]
Comment (AlexandruCihodaru) [reply src/allocation_engine/interval_tree.rs:532]: Exactly the unit tests are going to be changed when the insert method will be created

File: src/allocation_engine/interval_tree.rs [hunk 6]
```diff
@@ -185,8 +185,173 @@ struct InnerNode {
...
+
+    /// Updates cached information of the node.
+    #[allow(dead_code)]
+    fn update_cached_height(&mut self) {
+        self.height = max(height(&self.left), height(&self.right)) + 1;
+    }
+
+    /// Updates the sub-tree to keep balance.
+    #[allow(dead_code)]
+    fn updated_node(mut self: Box<Self>) -> Box<Self> {
```
Comment (studychao) [inline src/allocation_engine/interval_tree.rs:345]: not sure the meaning of `updated` here, maybe this should be `update_node`

File: src/allocation_engine/interval_tree.rs [hunk 7]
```diff
@@ -286,4 +451,158 @@ mod tests {
...
+        );
+        assert_eq!(
+            root_node.search_superset(&Range::new(0x1ff, 0x300).unwrap()),
+            None
+        );
+    }
+
+    #[test]
+    fn test_delete_root_balance_right() {
+        let mut unbalanced_tree =
```
Comment (studychao) [inline src/allocation_engine/interval_tree.rs:515]: I think `delete_root` is defined to be used on balanced_tree.
For example, if you add other larger node (e.g. `right_node5`) on the right subtree, the delete_root will panic because the height diffence during rotation is larger than 2.

File: src/allocation_engine/interval_tree.rs [hunk 8]
```diff
@@ -286,4 +451,158 @@ mod tests {
...
+
+    #[test]
+    fn test_search_superset() {
+        let left_child = InnerNode::new(Range::new(0x90, 0x99).unwrap(), NodeState::Free);
+        let right_child = InnerNode::new(Range::new(0x200, 0x2ff).unwrap(), NodeState::Free);
+        let mut root_node = InnerNode::new(Range::new(0x100, 0x110).unwrap(), NodeState::Allocated);
+        root_node.left = Some(Box::new(left_child));
+        root_node.right = Some(Box::new(right_child.clone()));
+
+        assert_eq!(
```
Comment (andreeaflorescu) [inline src/allocation_engine/interval_tree.rs:479]: Can you write a bit about the test cases here? There are a lot of asserts and it's a bit hard to follow.

File: src/allocation_engine/interval_tree.rs [hunk 9]
```diff
@@ -185,8 +185,173 @@ struct InnerNode {
...
+            2 => self.rotate_left_successor(),
+            -2 => self.rotate_right_successor(),
+            _ => unreachable!(),
+        }
+    }
+
+    /// Performs a single left rotation on this node.
+    #[allow(dead_code)]
+    fn rotate_left(mut self: Box<Self>) -> Box<Self> {
+        let mut new_root = self.right.take().expect("Node is broken");
```
Comment (andreeaflorescu) [inline src/allocation_engine/interval_tree.rs:251]: Why is this unwrap safe? Can you add a comment? We shouldn't have unwraps in the code unless they can be caused only by programming errors.

File: src/allocation_engine/interval_tree.rs (hunk 6 above)
Comment (andreeaflorescu) [reply src/allocation_engine/interval_tree.rs:345]: s/updated_node/update_node

File: src/allocation_engine/interval_tree.rs (hunk 7 above)
Comment (studychao) [reply src/allocation_engine/interval_tree.rs:530]: After discussion with Alex, `test_delete_root_*` are temporary unit tests for `delete_root` function and they will be replaced later when `insert` and `delete` for interval tree are introduced.

File: src/allocation_engine/interval_tree.rs (hunk 9 above)
Comment (andreeaflorescu) [reply src/allocation_engine/interval_tree.rs:255]: Reading the comments, I think it's actually safer (in the sense that it's less prone to mistakes in the future) if we just return an `Option` here, and do the unwrap in the function where we validated that the child exists. The risk is that someone makes this function public, and then it can be called without doing the previously mentioned validation.

Comment (AlexandruCihodaru) [reply src/allocation_engine/interval_tree.rs:255]: Okay. Will update it

**Review by studychao — APPROVED** [review]

**Review by gsserge — APPROVED** [review]
